warehouse/packaging/metadata_verification.py

```diff
@@ -41,6 +50,86 @@ def _verify_url_pypi(url: str, project_name: str, project_normalized_name: str)
     )
 
 
+def _get_url_content(url: rfc3986.api.URIReference, max_length_bytes: int) -> str:
+    """
+    Get at least `max_length_bytes` of the contents of a URL
+
+    Raises `requests.exception.RequestException` on any errors while trying to
+    access the URL.
+    """
+    headers = {"User-Agent": "warehouse/1.0 (PyPI URL verifier, https://pypi.org)"}
+    r = requests.get(
+        url.unsplit(),
+        allow_redirects=False,
+        timeout=5,
+        stream=True,
+        headers=headers,
+    )
+    r.raise_for_status()
+
+    content = next(r.iter_content(max_length_bytes))
+    return content
+
+
+def _verify_url_meta_tag(
+    url: str, project_name: str, project_normalized_name: str
+) -> bool:
+    user_uri = rfc3986.api.uri_reference(url).normalize()
+
+    # Require the presence of "https" and the host name. If a port
+    # is present, it must be 443.
+    validator = (
+        rfc3986_validators.Validator()
+        .require_presence_of("scheme", "host")
+        .allow_schemes("https")
+        .allow_ports("443")
+    )
+
+    try:
+        validator.validate(user_uri)
+    except rfc3986ValidationError:
+        return False
+
+    # Only allow registered names for the host (e.g.: google.com, api.github.com),
+    # don't allow IP addresses.
+    if IPv4_MATCHER.fullmatch(user_uri.host) or IPv6_MATCHER.fullmatch(user_uri.host):
+        return False
+
+    # The domain name should not resolve to a private or shared IP address
+    try:
+        address_tuples = socket.getaddrinfo(user_uri.host, user_uri.port)
+        for family, _, _, _, sockaddr in address_tuples:
+            ip_address: ipaddress.IPv4Address | ipaddress.IPv6Address | None = None
+            if family == socket.AF_INET:
+                ip_address = ipaddress.IPv4Address(sockaddr[0])
+            elif family == socket.AF_INET6:
+                ip_address = ipaddress.IPv6Address(sockaddr[0])
+            if ip_address is None or not ip_address.is_global:
+                return False
+    except (socket.gaierror, ipaddress.AddressValueError):
+        return False
+
+    # We get the first 1024 bytes
+    try:
+        content = _get_url_content(user_uri, max_length_bytes=1024)
+    except requests.exceptions.RequestException:
+        return False
```

Comment:
Flagging: this has an unfortunate TOC/TOU weakness, where `getaddrinfo` might return a public IP, and the subsequent resolution in `requests` might return a private/internal one. In other words, an attacker could race the check here to get it to pass.

I _think_ the correct way to do this is to either inspect the `socket` object underneath the `requests` response, _or_ to use the resolved IP directly and employ the `Host` header + SNI to resolve the correct domain from the server. But both of these are annoying to do 🙂 

For the latter, `HostHeaderSSLAdapter` from `requests_toolbelt` is possibly the most painless approach: https://toolbelt.readthedocs.io/en/latest/adapters.html#requests_toolbelt.adapters.host_header_ssl.HostHeaderSSLAdapter

(This may also be possible more easily via `urllib3` instead -- maybe @sethmlarson knows?)

Reply:
This *should* be easier with urllib3, you can send your own `Host` and SNI values more directly.

Reply:
Just confirmed with Wireshark that this works for urllib3:

```python
http = urllib3.HTTPSConnectionPool(
    host="93.184.215.14",
    port=443,
    headers={"Host": "example.com"},
    server_hostname="example.com",
    assert_hostname="example.com",
)
resp = http.request("GET", "/")
```

Sends SNI of `example.com`, asserts `example.com` on the cert, sends `example.com` in the `Host` header, but doesn't do any DNS resolution. I should probably add a super-strict integration test for this construction to the urllib3 test suite, the individual parts are tested but having it all tested together is quite nice to double-check.

Reply:
Amazing, thanks @sethmlarson!

Reply:
Thanks! The implementation now uses `urllib3`